### PR TITLE
Implement ChannelShuffle op with XNNPACK

### PR DESCRIPTION
**Pull request by taivu1998**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43602 Implement ChannelShuffle op with XNNPACK**
* #43509 Add benchmark for channel_shuffle operator

Differential Revision: [D23334952](https://our.internmc.facebook.com/intern/diff/D23334952)

### CODE REVIEW

File: aten/src/ATen/native/ChanelShuffle.cpp [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 #include <ATen/NamedTensorUtils.h>
 #include <ATen/NativeFunctions.h>
+#include <ATen/native/xnnpack/Engine.h>
```
Comment (kimishpatel) [inline aten/src/ATen/native/ChanelShuffle.cpp:5]: You can protect this also with USE_XNNPACK

File: aten/src/ATen/native/ChanelShuffle.cpp [hunk 2]
```diff
@@ -11,6 +12,13 @@ namespace at {
 namespace native {
 
 Tensor channel_shuffle(const Tensor& self, int64_t groups) {
+#if defined(USE_XNNPACK)
+  if (self.is_contiguous(MemoryFormat::ChannelsLast) &&
+      xnnpack::use_channel_shuffle(self, groups)) {
+    return xnnpack::channel_shuffle(self, groups);
+  }
+#endif
+
```
Comment (kimishpatel) [inline aten/src/ATen/native/ChanelShuffle.cpp:21]: Let's move all this after line 33 where all the checks are done. The checks should be common for both.

File: aten/src/ATen/native/xnnpack/ChannelShuffle.cpp [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+namespace xnnpack {
+
+bool use_channel_shuffle(
+    const Tensor& input,
+    const int64_t groups) {
+  using namespace internal;
+
+  // Here are the list of conditions required for this code path to be taken:
+  // * Input must be 4D CPU float tensor with no gradients.
+  // * output_max must be greater than output_min.
```
Comment (kimishpatel) [inline aten/src/ATen/native/xnnpack/ChannelShuffle.cpp:17]: What is the output_max comment? I dont see it being used anywhere here.

File: aten/src/ATen/native/xnnpack/ChannelShuffle.cpp [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+#include <ATen/native/utils/Factory.h>
+
+namespace at {
+namespace native {
+namespace xnnpack {
+
+bool use_channel_shuffle(
+    const Tensor& input,
+    const int64_t groups) {
+  using namespace internal;
```
Comment (kimishpatel) [inline aten/src/ATen/native/xnnpack/ChannelShuffle.cpp:13]: Why do you need this (`using namespace internal`)?

File: benchmarks/operator_benchmark/pt/channel_shuffle_test.py [hunk 5]
```diff
@@ -39,7 +39,7 @@ class ChannelSHuffleBenchmark(op_bench.TorchBenchmarkBase):
     def init(self, batch_size, channels_per_group, height, width, groups):
         channels = channels_per_group * groups
         data_shape = (batch_size, channels, height, width)
-        self.input_data = torch.rand(data_shape)
+        self.input_data = torch.rand(data_shape).contiguous(memory_format=torch.channels_last)
```
Comment (kimishpatel) [inline benchmarks/operator_benchmark/pt/channel_shuffle_test.py:42]: Let's benchmark both formats?

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Overall this looks good. Please benchmark for both channels first and channels last and report the numbers here.

File: aten/src/ATen/native/xnnpack/ChannelShuffle.cpp (hunk 3 above)
Comment (taivu1998) [reply aten/src/ATen/native/xnnpack/ChannelShuffle.cpp:17]: Sorry, I need to remove that.

File: aten/src/ATen/native/xnnpack/ChannelShuffle.cpp (hunk 4 above)
Comment (taivu1998) [reply aten/src/ATen/native/xnnpack/ChannelShuffle.cpp:13]: That's for using `Layout::Activation4D` in the function. I think `MaxPooling.cpp` did the same thing.

File: benchmarks/operator_benchmark/pt/channel_shuffle_test.py (hunk 5 above)
Comment (taivu1998) [reply benchmarks/operator_benchmark/pt/channel_shuffle_test.py:42]: Do you mean adding both formats for both diffs or just the top diff?

File: aten/src/ATen/native/xnnpack/ChannelShuffle.cpp (hunk 4 above)
Comment (kimishpatel) [reply aten/src/ATen/native/xnnpack/ChannelShuffle.cpp:13]: Oh I see. OK thats fine.

**Review by kimishpatel — COMMENTED** [review]
Looks good to me. Add benchmark results to see the perf improvement.

**Comment by taivu1998** [comment]
| Hyperparameters  | Channel First (original) | Channel Last (original) | Channel Last (XNNPACK) |
| ------------- | -------------: | -------------: | -------------: |
| Batch_size: 2, Channels per group: 16, Height: 16, Width: 16, Groups: 2  | 10.510 us | 19.407 us | 17.920 us |
| Batch_size: 2, Channels per group: 32, Height: 32, Width: 32, Groups: 2  | 46.724 us | 137.186 us | 65.069 us |
| Batch_size: 4, Channels per group: 32, Height: 32, Width: 32, Groups: 4  | 856.847 us | 1204.111 us | 690.858 us |
| Batch_size: 4, Channels per group: 64, Height: 64, Width: 64, Groups: 4  | 2808.228 us | 26120.442 us | 2765.157 us |
| Batch_size: 8, Channels per group: 64, Height: 64, Width: 64, Groups: 8  | 54180.293 us | 159411.737 us | 49908.312 us |
| Batch_size: 16, Channels per group: 64, Height: 64, Width: 64, Groups: 16  | 218493.526 us | 1107034.267 us | 200149.906 us |

**Comment by kimishpatel** [comment]
> Hyperparameters	Channel First (original)	Channel Last (original)	Channel Last (XNNPACK)
> Batch_size: 2, Channels per group: 16, Height: 16, Width: 16, Groups: 2	10.510 us	19.407 us	17.920 us
> Batch_size: 2, Channels per group: 32, Height: 32, Width: 32, Groups: 2	46.724 us	137.186 us	65.069 us
> Batch_size: 4, Channels per group: 32, Height: 32, Width: 32, Groups: 4	856.847 us	1204.111 us	690.858 us
> Batch_size: 4, Channels per group: 64, Height: 64, Width: 64, Groups: 4	2808.228 us	26120.442 us	2765.157 us
> Batch_size: 8, Channels per group: 64, Height: 64, Width: 64, Groups: 8	54180.293 us	159411.737 us	49908.312 us
> Batch_size: 16, Channels per group: 64, Height: 64, Width: 64, Groups: 16	218493.526 us	1107034.267 us	200149.906 us

This definitely is not as good as I expected, in some cases worse, but I think I know what is going on.

So XNNPACK can access input out of bounds. E.g. if input is of 64 bytes it is possible that XNNPACK will try to access up 64+16 bytes. So the call `allocate_padded_contiguous_if_needed` checks if the allocation is already padded (by checking if this tensor was allocated using `DefaultMobileCPUAllocator`). If it is not padded, it allocates new tensor and copies older one to new one. So most likely we are paying the cost of copying and hence the speedup is not as much.

In mobile however the case is different. In mobile builds we use `DefaultCPUMobileAllocator` which already does padded allocation. Thus we do not pay any additional cost. Let's do two things.

1) Locally change this line https://github.com/pytorch/pytorch/blob/master/c10/core/CPUAllocator.cpp#L239 to `return GetDefaultMobileCPUAllocator();` and rerun the experiments. We will not commit this change, but this is to confirm my intuition above.
2) Change all `if defined(USE_XNNPACK)` in ChanneShuffle.cpp to `#ifdef C10_MOBILE` so as to go through this path only on mobile. Make that change in this PR and we will commit with that.

**Comment by taivu1998** [comment]
> > Hyperparameters	Channel First (original)	Channel Last (original)	Channel Last (XNNPACK)
> > Batch_size: 2, Channels per group: 16, Height: 16, Width: 16, Groups: 2	10.510 us	19.407 us	17.920 us
> > Batch_size: 2, Channels per group: 32, Height: 32, Width: 32, Groups: 2	46.724 us	137.186 us	65.069 us
> > Batch_size: 4, Channels per group: 32, Height: 32, Width: 32, Groups: 4	856.847 us	1204.111 us	690.858 us
> > Batch_size: 4, Channels per group: 64, Height: 64, Width: 64, Groups: 4	2808.228 us	26120.442 us	2765.157 us
> > Batch_size: 8, Channels per group: 64, Height: 64, Width: 64, Groups: 8	54180.293 us	159411.737 us	49908.312 us
> > Batch_size: 16, Channels per group: 64, Height: 64, Width: 64, Groups: 16	218493.526 us	1107034.267 us	200149.906 us
> 
> This definitely is not as good as I expected, in some cases worse, but I think I know what is going on.
> 
> So XNNPACK can access input out of bounds. E.g. if input is of 64 bytes it is possible that XNNPACK will try to access up 64+16 bytes. So the call `allocate_padded_contiguous_if_needed` checks if the allocation is already padded (by checking if this tensor was allocated using `DefaultMobileCPUAllocator`). If it is not padded, it allocates new tensor and copies older one to new one. So most likely we are paying the cost of copying and hence the speedup is not as much.
> 
> In mobile however the case is different. In mobile builds we use `DefaultCPUMobileAllocator` which already does padded allocation. Thus we do not pay any additional cost. Let's do two things.
> 
> 1. Locally change this line https://github.com/pytorch/pytorch/blob/master/c10/core/CPUAllocator.cpp#L239 to `return GetDefaultMobileCPUAllocator();` and rerun the experiments. We will not commit this change, but this is to confirm my intuition above.
> 2. Change all `if defined(USE_XNNPACK)` in ChanneShuffle.cpp to `#ifdef C10_MOBILE` so as to go through this path only on mobile. Make that change in this PR and we will commit with that.

1. I tried making that change, and still got similar performance as above.
2. I'm wondering why we only want to enable it for mobile. Even though XNNPACK didn't outperform channel first performance, it still performed significantly better in the channel last case. Hence, we should always enable it in the channel last case, shouldn't we?

**Comment by kimishpatel** [comment]
> 1. I tried making that change, and still got similar performance as above.

Can you paste the local diff here? If it is still giving similar perf, then I dont think that necessarily bodes well. We may have to look at the profile and understand why it performs worse. In that case we may just temporarily disable it.

> 2. I'm wondering why we only want to enable it for mobile. Even though XNNPACK didn't outperform channel first performance, it still performed significantly better in the channel last case. Hence, we should always enable it in the channel last case, shouldn't we?

More often than not, channels last is used only for mobile, as far as I know, plus less surprises in performance for non-mobile case.
Furthermore channels last performs worse here because the default implementation is better for channels first. Thus if we use that implementation for channels last we are creating a baseline which is much worse.

Come to think of it, it might be the case that channel shuffle is better done in NCHW than NHWC. So let's stick with mobile use case plan.

**Comment by taivu1998** [comment]
So, as @kimishpatel suggested above, the XNNPACK channel shuffle may perform better in the mobile use case, because we don't need to pay the cost of allocating and copying tensors. To test this, I modified https://github.com/pytorch/pytorch/blob/master/c10/core/CPUAllocator.cpp#L225-L244 as follows to test mobile case.

```
1	
diff --git a/c10/core/CPUAllocator.cpp b/c10/core/CPUAllocator.cpp
index 24e92982b8..f004dffe33 100644
--- a/c10/core/CPUAllocator.cpp
+++ b/c10/core/CPUAllocator.cpp
@@ -222,7 +222,7 @@ at::Allocator* GetDefaultMobileCPUAllocator() {
   return &g_mobile_cpu_allocator;
 }
 
-#ifdef C10_MOBILE
+// #ifdef C10_MOBILE
 
 at::Allocator* GetDefaultCPUAllocator() {
   return GetDefaultMobileCPUAllocator();
@@ -230,19 +230,19 @@ at::Allocator* GetDefaultCPUAllocator() {
 
 REGISTER_ALLOCATOR(DeviceType::CPU, &g_mobile_cpu_allocator);
 
-#else
+// #else
 
-// Global default CPU Allocator
-static DefaultCPUAllocator g_cpu_alloc;
+// // Global default CPU Allocator
+// static DefaultCPUAllocator g_cpu_alloc;
 
-at::Allocator* GetDefaultCPUAllocator() {
-  // return &g_cpu_alloc;
-  return GetDefaultMobileCPUAllocator();
-}
+// at::Allocator* GetDefaultCPUAllocator() {
+//   // return &g_cpu_alloc;
+//   return GetDefaultMobileCPUAllocator();
+// }
 
-REGISTER_ALLOCATOR(DeviceType::CPU, &g_cpu_alloc);
+// REGISTER_ALLOCATOR(DeviceType::CPU, &g_mobile_cpu_allocator);
 
-#endif /* C10_Mobile */
+// #endif /* C10_Mobile */
 
 void ProfiledCPUMemoryReporter::New(void* ptr, size_t nbytes) {
   if (nbytes == 0) {
```

Below is the performance comparison.

| Hyperparameters  | Channel First (original) | Channel Last (XNNPACK) | Improvement |
| ------------- | -------------: | -------------: | -------------: |
| Batch_size: 2, Channels per group: 16, Height: 16, Width: 16, Groups: 2  | 10.242 us | 16.651 us | -62.6% |
| Batch_size: 2, Channels per group: 32, Height: 32, Width: 32, Groups: 2  | 45.091 us | 12.119 us | 73.1% |
| Batch_size: 4, Channels per group: 32, Height: 32, Width: 32, Groups: 4  | 197.602 us | 14.119 us | 92.9% |
| Batch_size: 4, Channels per group: 64, Height: 64, Width: 64, Groups: 4  | 3069.089 us | 38.133 us | 98.8% |
| Batch_size: 8, Channels per group: 64, Height: 64, Width: 64, Groups: 8  | 53603.974 us | 91.474 us | 99.8% |
| Batch_size: 16, Channels per group: 64, Height: 64, Width: 64, Groups: 16  | 190912.903 us | 120.748 us | 99.9% |

So the average improvement here is about 70%.

**Comment by taivu1998** [comment]
In addition, I tried switching to `#if defined(USE_XNNPACK)` and then ran the unit test `python test/test_nn.py TestNN.test_channel_shuffle`. The program successfully passed the test.

**Review by kimishpatel — APPROVED** [review]
Make sure all tests/builds are passing.
